twispay.php

```diff
@@ -243,8 +328,80 @@ public function hookDisplayBackOfficeHeader()
             $this->context->controller->addJS($this->_path.'views/js/back.js');
             $this->context->controller->addCSS($this->_path.'views/css/back.css');
         }
+        if ($this->context->cookie->redirect_error) {
+            /** Display persistent error */
+            $this->context->controller->errors[] = $this->context->cookie->redirect_error;
+            /** Clean persistent error */
+            unset($this->context->cookie->redirect_error);
+        }
+    }
+
+    /** Display last order information on Admin > Orders > Order */
+    public function hookDisplayAdminOrderLeft($params)
+    {
+        $id_order = (int)$params['id_order'];
+        $data = Db::getInstance()->getRow('SELECT * FROM `'._DB_PREFIX_.'twispay_transactions`
+        WHERE `id_cart` = (SELECT `id_cart` FROM `'._DB_PREFIX_.'orders` WHERE `id_order` = "'.$id_order.'") ORDER BY `date` DESC');
+        if (!$data) {
+            return false;
+        } else {
+            /** Link the info to query */
+            return $this->buildOrderMessage($data);
+        }
+    }
+
+    /** Load template with assigned data */
+    public function buildOrderMessage($data)
+    {
+        $this->context->smarty->assign('data', $data);
+        return $this->display(__FILE__, 'views/templates/admin/payment_message.tpl');
+    }
+
+    /**
+    * Create Front interface
+    */
+    /** Hook for order status update action */
+    public function hookActionOrderStatusUpdate($params)
+    {
+        $status = $params['newOrderStatus']?$params['newOrderStatus']->id:false;
+
+        if ($status === 7/** Refund status id */) {
+            /** If transaction is registred in twispay transactions list */
+            $transaction = Twispay_Transactions::getTransactionByCartId($params['cart']->id);
+            if ($transaction) {
+                if ($transaction['status'] == Twispay_Status_Updater::$RESULT_STATUSES['REFUND_OK']) {
+                    Twispay_Logger::api_log($this->l('Order already refunded.'));
+                    $this->context->cookie->redirect_error = $this->l('Order already refunded.');
+                } else {
+                    $keys = self::getKeysInfo();
+                    if (!$keys) {
+                        $this->context->cookie->redirect_error = $this->l('Twispay refund error: ').$this->l('Invalid API Keys.');
+                        /** Redirect to order page */
+                        /** Skip the part when the status is set */
+                        Tools::redirect($_SERVER['HTTP_REFERER']);
+                        die();
+                    }
+                    $refund = Twispay_Transactions::refundTransaction($transaction, $keys);
```

Comment:
There is a missing parameter `$module` from the method call.

Reply:
Done